### ISSUE TITLE
MIDI-MAP CPU usage can get quite high

### ISSUE THREAD

**Issue by MaZe6K** [VCVRack/Rack#1473]
### Details

Operating system: Windows 7
Rack version: v1.1.1
Hardware relevant to your issue (e.g. graphic card model, audio/MIDI device):
I use two LAUNCHCONTROL XL and an AKAI MIDIMIX and want to map a lot.

### Summary

I don't know if this is actually a bug or if it can be avoided, the thing is:
When I start mapping, the CPU usage of the MAP module increases with every mapping. But when I save the project and load it again, the CPU usage is very low, until I start to use the controls - with every use of a control CPU usage gets up again. Since it is low when freshly loaded, I hope that it could be made to stay low.
In my use case I need 3 MIDI MAP modules, and the CPU usage quickly goes up to 5% per module, which can be a limiting factor for me.
I hope this can be made to stay low, if not, sorry to bother.

### Steps to reproduce

map a lot of controls, save the project, load it up again, start using the controls, watch the CPU usage.

**Comment by MaZe6K** [comment]
![new mashine 2+ CPU](https://user-images.githubusercontent.com/50169405/61766899-c3c20f80-ade2-11e9-839c-25463778c294.jpg)

this is with an i7 and 3 threads activated... currently the midi mapping has become my biggest performance issue

**Comment by Coirt** [comment]
You've got 100+ controls mapped and that's only what is shown in the screenshot. 

What is your sample and engine rate set to? What audio interface are you using?

**Comment by MaZe6K** [comment]
I run it at 44.1kHz on a RME Multiface II with a buffer of 512 samples, the buffer size does not seem to make a difference, I can run it at 256 or 1024 with the same performance.
I have.. let's see... I think 162 controls mapped on two Launch Controls, a Midimix and a few on a BSP... it's a lot but I need all of them, because I want to go full on live machine with it.
Maybe the MIDI-MAP module runs at sample rate and could be clocked down to 1000Hz, which would still be fast enough? Don't know, but I think I can use MIDI automation without a noticeable overhead in other software (DAWs or Reaktor), so I wonder if this is neccessary.

**Comment by Coirt** [comment]
Does the engine sample rate match your audio sample rate?

**Comment by MaZe6K** [comment]
yes, just edited my answer... same rate

**Comment by MaZe6K** [comment]
Or maby the MIDI-MAP module could reset itself every few seconds to the state that it's in when freshly loaded? When I load the patch CPU use per module is around 0.4%, with every control used it goes up, when all controls have been used once it stays like on the screenshot. Does the first move of a knob have a higher latency then when I move it a second time or something?

**Comment by Coirt** [comment]
What are all block sizes of the Host VST's set to? Right click on host to see.

**Comment by Coirt** [comment]
I can see from the screenshot that some devices have cables in them are these devices set up for polyphony?

Do you have them set to 16 channels? does decreasing channels change the usage?

**Comment by MaZe6K** [comment]
I run the VSTs with a block size of 32 samples, one VST is set up for 4 channel polyphony, Reaktor (in Host XL) is my drumsynth and can be triggered by gates into the audio inputs...
But it all doesn't make a difference... I get the same high CPU consumption on the MIDI-MAP modules without any VSTs... and the patch is not tooo crazy... 3 simple synth voices, 1 VSTi with a 4 channel looper and additional filter, the drums in HOST XL, a reverb, delay and Solid Bus Compressor on the master... And I have this neat thing going on with 4 loopers, which can record the movement of the cuttoff of the 4 filters in the patch...
this is the whole patch:
![new mashine 2+](https://user-images.githubusercontent.com/50169405/61824523-813e1880-ae5e-11e9-9022-ad3e860f4bbb.jpg)
But this all doesn't matter, I allways get this behviour with the MIDI-MAP modules, regardless of Buffer sizes and VST plugins, the number of used mappings is what causes it.

**Comment by stoermelder** [comment]
> this is with an i7 and 3 threads activated... currently the midi mapping has become my biggest performance issue

I'm working on a module called MIDI-CAT with some additional features compared to MIDI-MAP. I also did some optimization on cpu usage. If you like you can get the latest build of my plugin and check if you get better performance.

**Comment by Coirt** [comment]
I found that if a VST uses a different block size than the audio block size it can be a hindrance on the chain. Try matching the audio buffer and see if this makes any difference, try find a balance. I know you will need the smallest buffer possible for the setup.

**Comment by MaZe6K** [comment]
@stoermelder yes, please, i would love that, can you give me a link?
@Coirt I don't think that it has anything to do with the rest of the patch and buffer sizes, I allways get this high CPU usage, even if no VSTs are present in the patch.

**Comment by Coirt** [comment]
Have you tried it?

**Comment by stoermelder** [comment]
I don’t want post a link here.
Please go on my GitHub repo for PackOne, click Releases and scroll to the bottom. There you find „nightly builds“ of my code.

**Comment by MaZe6K** [comment]
@coirt: just to really make sure I have just deleted all VSTs in the patch and I have the same CPU usage (around 9% per fully mapped controller)
@stoermelder ok, cool, I will check it out, thank you!

**Comment by stoermelder** [comment]
After reading your issue yesterday I added an option to import presets from MIDI-MAP on the context menu. This makes it quite easy and fast to test it if MIDI-MAP is already set up.

**Comment by Coirt** [comment]
@MaZe6K is that with restarting rack. Could you try to change the buffer sizes on the vst's to match racks buffer, close rack and open it again.

**Comment by MaZe6K** [comment]
@stoermelder I just checked it out, and it's much better. With the same fully mapped controller I get around 3-4% CPU usage with it. This might help me a lot with what I want to do, Thank you very much!
Importing presets is almost a bit of overkill :), but nice, I will check that out as well.

**Comment by Coirt** [comment]
@MaZe6K right, but could you confirm that you've tried it?

**Comment by MaZe6K** [comment]
@Coirt yes. just to make really really really sure I have just done that, same CPU usage.

**Comment by MaZe6K** [comment]
@coirt Just to test that again I also set all block sizes to the asio buffer size (512 samples) and did a restart, with the same cpu usage also.

I then activated a 4th thread, and to my surprise the CPU usage meter went further up, up to around 12%, so multi-threading doen's help it one bit neither.

@stoermelder another test didn't show the same extreme results, one module still went up to around 6%, (still on 3 threads), but generally they seem to roughly halve the CPU consumption for MIDI mapping.
The VCV modules are very low in usage when freshly loaded and then go up with each used control, your modules start with their full usage and stay there.
I also like that they allow me to still use the mouse - and bidirectional mapping, nice! MIDI-MAP preset loading also works.
Could they use even less if mousecontrol and/or bidirectional mapping would be deactivated? That would be awesome... a very nice addition anyway, I'm really happy about this.

**Comment by stoermelder** [comment]
I had an idea today and was able to reduce the cpu consumption to almost zero, just processing actual midi events will take more cpu. Feel free to test the latest build.

**Comment by MaZe6K** [comment]
YESSS!!! I just did that, CPU consumption is around 0.5% now and only goes up slightly when moving knobs. My eyes just watered a little bit when testing. This is so beautiful!

**Comment by MaZe6K** [comment]
@stoermelder Now I can't help but to ask you what you might think of this: https://github.com/VCVRack/Rack/issues/1424
